nav2_rrtconnect_planner/src/rrt_connect.cpp

```diff
@@ -0,0 +1,269 @@
+{
+  RCLCPP_INFO(
+    node_->get_logger(), "Deactivating plugin %s of type NavfnPlanner",
+    name_.c_str());
+}
+
+void RRTConnect::initializeOMPL()
+{
+  // Defining Bounds
+  bounds_ = new ompl::base::RealVectorBounds(2);
+  bounds_->setLow(0, costmap_->getOriginX());
+  bounds_->setHigh(
+    0, costmap_->getSizeInMetersX() +
+    costmap_->getOriginX());
+  bounds_->setLow(1, costmap_->getOriginY());
+  bounds_->setHigh(
+    1, costmap_->getSizeInMetersY() +
+    costmap_->getOriginY());
+
+  RCLCPP_INFO(
+    node_->get_logger(),
+    "bounds: %f %f %f %f", bounds_->low[0], bounds_->low[1],
+    bounds_->high[0], bounds_->high[1]);
+
+  ompl_state_space_ = std::make_shared<ompl::base::SE2StateSpace>();
+  // Bounding the R2 state space
+  ompl_state_space_->as<ompl::base::SE2StateSpace>()->setBounds(*bounds_);
+
+  // Simple setup
+  ss_.reset(new ompl::geometric::SimpleSetup(ompl_state_space_));
+  // Setting state validity checker
+  ss_->setStateValidityChecker(
+    [this](const ompl::base::State * state) {
+      return this->isStateValid(state);
+    });
+  // Setting the state validity checking resolution
+  ss_->getSpaceInformation()->setStateValidityCheckingResolution(collision_checking_resolution_);
+
+  ompl_initialized_ = true;
+}
+
+nav_msgs::msg::Path RRTConnect::createPlan(
+  const geometry_msgs::msg::PoseStamped & start,
+  const geometry_msgs::msg::PoseStamped & goal)
+{
+  if (!ompl_initialized_) {
+    RCLCPP_INFO(node_->get_logger(), "Initializing OMPL!");
+    initializeOMPL();
+  }
+  // Returning variable
+  nav_msgs::msg::Path global_path;
+
+  // Checking if the goal and start state is in the global frame
+  if (start.header.frame_id != global_frame_) {
+    RCLCPP_ERROR(
+      node_->get_logger(), "Planner will only except start position from %s frame",
+      global_frame_.c_str());
+    return global_path;
+  }
+
+  if (goal.header.frame_id != global_frame_) {
+    RCLCPP_INFO(
+      node_->get_logger(), "Planner will only except goal position from %s frame",
+      global_frame_.c_str());
+    return global_path;
+  }
+
+  ompl::base::ScopedState<> ss_start(ompl_state_space_), ss_goal(ompl_state_space_);
+  ss_start = std::vector<double>{start.pose.position.x, start.pose.position.y};
+  ss_goal = std::vector<double>{goal.pose.position.x, goal.pose.position.y};
+  // Setting state and goal state
+  ss_->setStartAndGoalStates(ss_start, ss_goal);
+  // Assigning RRTConnect
+  ss_->setPlanner(
+    ompl::base::PlannerPtr(
+      new ompl::geometric::RRTConnect(
+        ss_->getSpaceInformation())));
+  // Load settings
+  ss_->setup();
+  // Counter for replanning
+  int counter = 0;
+  // Planning
+  while (!ss_->solve(solve_time_) && counter < max_trials_) {
+    ++counter;
+  }
+  // Plan found
+  if (counter < max_trials_) {
+    RCLCPP_INFO(
+      node_->get_logger(), "Found %d solutions",
+      ss_->getProblemDefinition()->getSolutionCount());
+    // Path simplification
+    ss_->simplifySolution();
```

Comment:
This maybe out of the scope of a simple tutorial but choosing between returning a simplified/original solution could be a parameter.